src/main/java/de/tum/in/www1/artemis/service/UserService.java

```diff
@@ -347,33 +369,52 @@ public void updateUser(String firstName, String lastName, String email, String l
-            user.setEmail(userDTO.getEmail().toLowerCase());
-            user.setImageUrl(userDTO.getImageUrl());
-            user.setActivated(userDTO.isActivated());
-            user.setLangKey(userDTO.getLangKey());
-            Set<Authority> managedAuthorities = user.getAuthorities();
-            managedAuthorities.clear();
-            userDTO.getAuthorities().stream().map(authorityRepository::findById).filter(Optional::isPresent).map(Optional::get).forEach(managedAuthorities::add);
-            this.clearUserCaches(user);
-            log.debug("Changed Information for User: {}", user);
-            return user;
-        }).map(UserDTO::new);
+    public User updateUser(User user, ManagedUserVM updatedUserDTO) {
+        final var oldGroups = user.getGroups();
+        this.clearUserCaches(user);
+        user.setLogin(updatedUserDTO.getLogin().toLowerCase());
+        user.setFirstName(updatedUserDTO.getFirstName());
+        user.setLastName(updatedUserDTO.getLastName());
+        user.setEmail(updatedUserDTO.getEmail().toLowerCase());
+        user.setImageUrl(updatedUserDTO.getImageUrl());
+        user.setActivated(updatedUserDTO.isActivated());
+        user.setLangKey(updatedUserDTO.getLangKey());
+        user.setGroups(updatedUserDTO.getGroups());
+        if (updatedUserDTO.getPassword() != null) {
+            user.setPassword(passwordEncoder().encode(updatedUserDTO.getPassword()));
+        }
+        Set<Authority> managedAuthorities = user.getAuthorities();
+        managedAuthorities.clear();
+        updatedUserDTO.getAuthorities().stream().map(authorityRepository::findById).filter(Optional::isPresent).map(Optional::get).forEach(managedAuthorities::add);
+        user = userRepository.save(user);
+        this.clearUserCaches(user);
+
+        forwardUserUpdates(user, oldGroups);
+
+        log.debug("Changed Information for User: {}", user);
+        return user;
+    }
+
+    private void forwardUserUpdates(User user, Set<String> oldGroups) {
+        final var updatedGroups = user.getGroups();
+        final var removedGroups = oldGroups.stream().filter(group -> !updatedGroups.contains(group)).collect(Collectors.toSet());
+        final var addedGroups = updatedGroups.stream().filter(group -> !oldGroups.contains(group)).collect(Collectors.toSet());
+        final var login = user.getLogin();
+        optionalVcsUserManagementService.ifPresent(vcsUserManagementService -> vcsUserManagementService.updateUser(user, removedGroups, addedGroups));
+        removedGroups.forEach(group -> artemisAuthenticationProvider.removeUserFromGroup(login, group));
+        addedGroups.forEach(group -> artemisAuthenticationProvider.addUserToGroup(login, group));
     }
 
     /**
      * Delete user based on login string
      * @param login user login string
      */
     public void deleteUser(String login) {
+        optionalVcsUserManagementService.ifPresent(userManagementService -> userManagementService.deleteUser(login));
```

Comment:
This should not delete a user in the external user management system (e.g. JIRA).

Reply:
I guess some comments would be good here as well

Reply:
This only deletes the user in the VCS and only if the bean is present (which is the case for GitLab, but not for Bitbucket). I also added some comments